Assets/MRTK/Core/Providers/InputAnimation/InputAnimation.cs

```diff
@@ -34,25 +34,123 @@ public class InputAnimationMarker
-            readonly public AnimationCurve RotationZ = new AnimationCurve();
-            readonly public AnimationCurve RotationW = new AnimationCurve();
+            public AnimationCurve PositionX = new AnimationCurve();
+            public AnimationCurve PositionY = new AnimationCurve();
+            public AnimationCurve PositionZ = new AnimationCurve();
+            public AnimationCurve RotationX = new AnimationCurve();
+            public AnimationCurve RotationY = new AnimationCurve();
+            public AnimationCurve RotationZ = new AnimationCurve();
+            public AnimationCurve RotationW = new AnimationCurve();
+
+            public void AddKey(float time, MixedRealityPose pose)
+            {
+                AddFloatKey(PositionX, time, pose.Position.x);
+                AddFloatKey(PositionY, time, pose.Position.y);
+                AddFloatKey(PositionZ, time, pose.Position.z);
+
+                AddFloatKey(RotationX, time, pose.Rotation.x);
+                AddFloatKey(RotationY, time, pose.Rotation.y);
+                AddFloatKey(RotationZ, time, pose.Rotation.z);
+                AddFloatKey(RotationW, time, pose.Rotation.w);
+            }
+            
+            /// <summary>
+            /// Optimizes the set of curves.
+            /// </summary>
+            /// <param name="positionThreshold">The maximum permitted error between the positions of the old and new curves, in units.</param>
+            /// <param name="rotationThreshold">The maximum permitted error between the rotations of the old and new curves, in degrees.</param>
+            /// <param name="partitionSize">The size of the partitions of the curves that will be optimized independently. Larger values will optimize the curves better, but may take longer.</param>
+            public void Optimize(float positionThreshold, float rotationThreshold, int partitionSize)
+            {
+                OptimizePositionCurve(ref PositionX, ref PositionY, ref PositionZ, positionThreshold, partitionSize);
+                OptimizeRotationCurve(ref RotationX, ref RotationY, ref RotationZ, ref RotationW, rotationThreshold, partitionSize);
+            }
+
+            public MixedRealityPose Evaluate(float time)
+            {
+                float px = PositionX.Evaluate(time);
+                float py = PositionY.Evaluate(time);
+                float pz = PositionZ.Evaluate(time);
+                float rx = RotationX.Evaluate(time);
+                float ry = RotationY.Evaluate(time);
+                float rz = RotationZ.Evaluate(time);
+                float rw = RotationW.Evaluate(time);
+
+                var pose = new MixedRealityPose();
+                
+                pose.Position = new Vector3(px, py, pz);
+                pose.Rotation = new Quaternion(rx, ry, rz, rw);
+                pose.Rotation.Normalize();
+                
+                return pose;
+            }
+        }
+        
+        public class RayCurves
```

Comment:
Please add summary comments for public classes, also I would suggest moving this to another file.